packages/microeditor-envelope/src/CompositeEditorFactory.ts

```
@@ -0,0 +1,44 @@
+/*
+ * Copyright 2019 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as AppFormer from "@kogito-tooling/core-api";
+import { LanguageData } from "@kogito-tooling/core-api";
+import { EnvelopeBusInnerMessageHandler } from "./EnvelopeBusInnerMessageHandler";
+import { EditorFactory } from "./EditorFactory";
+
+/**
+ * Composite Factory of Editors to be created inside the envelope. This implementation delegates potential construction
+ * to specific concrete factories. The first concrete factory to return a promise to create an editor is selected.
+ */
+export class CompositeEditorFactory implements EditorFactory<LanguageData> {
+  private factories: Array<EditorFactory<LanguageData>> = new Array<EditorFactory<LanguageData>>();
+
+  constructor(factories: Array<EditorFactory<LanguageData>>) {
+    this.factories = factories;
+  }
+
+  public supports(languageData: LanguageData) {
+    return this.factories.filter(f => f.supports(languageData)).length > 0;
+  }
+
+  public createEditor(
+    languageData: LanguageData,
+    messageBus: EnvelopeBusInnerMessageHandler
+  ): Promise<AppFormer.Editor> {
+    const candidates = this.factories.filter(f => f.supports(languageData));
+    return candidates[0].createEditor(languageData, messageBus);
```

Can you throw an error if `candidates.length !== 1`?

---

Sure thing done... and test added!